File: nvpl_blas/c/zher2k.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_zher2k
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t K = 2;
    nvpl_int_t lda = 2;
    nvpl_int_t ldb = 2;
    nvpl_int_t ldc = 2;
    nvpl_dcomplex_t alpha = {1.0f, 2.0f};
    double beta = 1.0f;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE trans = CblasNoTrans;
    enum CBLAS_UPLO uplo = CblasUpper;
    nvpl_dcomplex_t * A = NULL;
    nvpl_dcomplex_t * B = NULL;
    nvpl_dcomplex_t * C = NULL;

    printf("\nExample: cblas_zher2k for the Hermitian rank-2k update\n\n");
    printf("#### args: n=%" PRId64 ", k=%" PRId64 ", lda=%" PRId64 ", ldb=%" PRId64 ", ldc=%" PRId64 ", "
            "trans=%c, uplo=%c, order=%c\n", (int64_t)N, (int64_t)K, (int64_t)lda, (int64_t)ldb,
            (int64_t)ldc, transpose_to_char(trans), uplo_to_char(uplo), order_to_char(order));
    printf("           alpha=(%g, %g), beta=%g\n", alpha.real, alpha.imag, beta);

    nvpl_int_t rowsA = (trans == CblasNoTrans) ? N : K;
    nvpl_int_t colsA = (trans == CblasNoTrans) ? K : N;
    nvpl_int_t rowsB = (trans == CblasNoTrans) ? N : K;
    nvpl_int_t colsB = (trans == CblasNoTrans) ? K : N;
    nvpl_int_t rowsC = N;
    nvpl_int_t colsC = N;

    // allocate memory
    A = (nvpl_dcomplex_t *)malloc(lda * colsA  * sizeof(nvpl_dcomplex_t));
    B = (nvpl_dcomplex_t *)malloc(ldb * colsB  * sizeof(nvpl_dcomplex_t));
    C = (nvpl_dcomplex_t *)malloc(ldc * colsC  * sizeof(nvpl_dcomplex_t));

    // fill data
    fill_zmatrix(A, rowsA, colsA, lda, order, Full, CblasNonUnit);
    fill_zmatrix(B, rowsB, colsB, ldb, order, Full, CblasNonUnit);
    fill_zmatrix(C, rowsC, colsC, ldc, order, UPLO_TO_FILL_MODE(uplo), CblasNonUnit);

    // print input data
    print_zmatrix(A, rowsA, colsA, lda, "A", order);
    print_zmatrix(B, rowsB, colsB, ldb, "B", order);
    print_zmatrix(C, rowsC, colsC, ldc, "C", order);

    // call cblas_zher2k
    cblas_zher2k(order, uplo, trans, N, K, &alpha, A, lda, B, ldb, beta, C, ldc);

    // print result
    printf("\nAfter zher2k operation:\n");
    print_zmatrix(C, rowsC, colsC, ldc, "C", order);

    // release memory
    free(A);
    free(B);
    free(C);
    return 0;
}

File: nvpl_blas/f77/zher2k.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     ZHER2K
!
!******************************************************************************
      program   ZHER2K_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

      character*1      trans, uplo
      parameter        (trans='N', uplo='U')
      integer          n, k, i, j
      integer          lda, ldb, ldc
      parameter        (n=2, k=2,
     &                  lda=2, ldb=2, ldc=2)
      double complex   alpha
      double precision beta
      parameter        (alpha=(1.0,2.0), beta=1.0)
      double complex   a(lda,k), b(ldb,n), c(ldc,n)

*       External Subroutines
      external         print_zmatrix, fill_zmatrix
*
*      Executable Statements
*
      print*
      print 99, 'Example: ZHER2K for the Hermitian rank-2k update'
*
*       Print input data
      print*
      print 100, n, k, lda, ldb, ldc, trans, uplo
      print 101, alpha, beta
      call fill_zmatrix(a,n,k,lda,'F','N')
      call fill_zmatrix(b,n,k,ldb,'F','N')
      call fill_zmatrix(c,n,n,ldc,uplo,'N')

      call print_zmatrix(a,n,k,lda,'A')
      call print_zmatrix(b,n,k,ldb,'B')
      call print_zmatrix(c,n,n,ldc,'C')
*
*      Call ZHER2K subroutine
      call ZHER2K(uplo,trans,n,k,alpha,a,lda,b,ldb,beta,c,ldc)
*
      print*
      print 102, 'After ZHER2K operation:'
      call print_zmatrix(c,n,n,ldc,'C')

      stop
 99   format(A)
 100  format('#### args: n=',i1,', k=',i1,', lda=',i1,
     &       ', ldb=',i1,', ldc=',i1,', trans=',a1,', uplo=',a1)
 101  format(11x,'alpha=(',f4.2,', ',f4.2,')',
     &       ', beta=',f4.2)
 102  format(A)
      end
